src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -17,28 +17,36 @@ namespace System.Text.Json
     public sealed partial class JsonSerializerOptions
     {
         // The global list of built-in simple converters.
-        private static readonly Dictionary<Type, JsonConverter> s_defaultSimpleConverters = GetDefaultSimpleConverters();
+        private static Dictionary<Type, JsonConverter>? s_defaultSimpleConverters;
 
         // The global list of built-in converters that override CanConvert().
-        private static readonly JsonConverter[] s_defaultFactoryConverters = new JsonConverter[]
-        {
-            // Check for disallowed types.
-            new DisallowedTypeConverterFactory(),
-            // Nullable converter should always be next since it forwards to any nullable type.
-            new NullableConverterFactory(),
-            new EnumConverterFactory(),
-            new JsonNodeConverterFactory(),
-            // IAsyncEnumerable takes precedence over IEnumerable.
-            new IAsyncEnumerableConverterFactory(),
-            // IEnumerable should always be second to last since they can convert any IEnumerable.
-            new IEnumerableConverterFactory(),
-            // Object should always be last since it converts any type.
-            new ObjectConverterFactory()
-        };
+        private static JsonConverter[]? s_defaultFactoryConverters;
 
         // The cached converters (custom or built-in).
         private readonly ConcurrentDictionary<Type, JsonConverter?> _converters = new ConcurrentDictionary<Type, JsonConverter?>();
 
+        internal void RootBuiltInConvertersAndTypeInfoCreator()
+        {
+            s_defaultSimpleConverters ??= GetDefaultSimpleConverters();
```

Comment:
Curious what necessitates this delayed initialization? I couldn't find either of these statics being initialized elsewhere. Having to call this method or risk getting nullref's feel too error prone to me.

If the issue is avoiding allocations in scenaria where they are not needed why not use lazily instantiated properties instead?

Reply:
It's not about delayed initialization or allocations. It's about allowing converters that are not needed in user apps to be trimmed out when ILLinker runs, and about allowing reflection based code-paths to be trimmed out since we are pre-generating type metadata. Refactoring the initialization this way allows the code to only be rooted when existing options/reflection-based methods of the serializer are used.